src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctionsTest.java

```diff
@@ -0,0 +1,275 @@
+
+    @Test
+    public void firstNameCombinationsForThreeNamesShouldReturnCombinations() {
+        List<String> firstNames = asList("John", "David", "Greg");
+        List<CandidateName> expected = asList(
+                new CandidateName("John", "David"),
+                new CandidateName("John", "Greg"),
+                new CandidateName("David", "John"),
+                new CandidateName("David", "Greg"),
+                new CandidateName("Greg", "John"),
+                new CandidateName("Greg", "David")
+        );
+
+        assertThat(firstNameCombinations(firstNames)).isEqualTo(expected);
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsShouldReturnEmptyListForEmptyInput() {
+        InputNames emptyInputNames = new InputNames(emptyList(), emptyList());
+        assertThat(nonAliasSurnameAsFirstNameCombinations(emptyInputNames)).isEqualTo(emptyList());
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsShouldReturnEmptyListForSingleInputName() {
+        InputNames oneFirstNameOnly = new InputNames(singletonList("somename"), emptyList());
+        InputNames oneLastNameOnly = new InputNames(emptyList(), singletonList("somename"));
+
+        assertThat(nonAliasSurnameAsFirstNameCombinations(oneFirstNameOnly)).isEqualTo(emptyList());
+        assertThat(nonAliasSurnameAsFirstNameCombinations(oneLastNameOnly)).isEqualTo(emptyList());
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsForOneFirstNameOneLastName() {
+        InputNames inputNames = new InputNames("John", "Smith");
+        List<CandidateName> expected = singletonList(new CandidateName("Smith", "John"));
+
+        assertThat(nonAliasSurnameAsFirstNameCombinations(inputNames)).isEqualTo(expected);
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsForTwoFirstNamesOneLastName() {
+        InputNames inputNames = new InputNames("John David", "Smith");
+        List<CandidateName> expected = asList(
+                new CandidateName("Smith", "John"),
+                new CandidateName("Smith", "David")
+        );
+
+        assertThat(nonAliasSurnameAsFirstNameCombinations(inputNames)).isEqualTo(expected);
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsForOneFirstNameTwoLastNames() {
+        InputNames inputNames = new InputNames("John", "Smith Evans");
+        List<CandidateName> expected = asList(
+                new CandidateName("Smith", "John"),
+                new CandidateName("Smith", "Evans"),
+                new CandidateName("Evans", "John"),
+                new CandidateName("Evans", "Smith")
+        );
+
+        assertThat(nonAliasSurnameAsFirstNameCombinations(inputNames)).isEqualTo(expected);
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsForTwoFirstNamesTwoLastNames() {
+        InputNames inputNames = new InputNames("John David", "Smith Evans");
+        List<CandidateName> expected = asList(
```

Comment:
Though I don't understand the alias rules, it feels like this class should do nothing if there are no alias names?

Reply:
The AliasSurnameCombinations does do nothing when there's no alias names. But it creates some combinations that don't involve alias names using some of the methods in its functions class.